### PR TITLE
AP_Logger: scripting log format allocation fix

### PR DESCRIPTION
**Pull request by IamPete1**
This fixes this valgrind issue, https://github.com/ArduPilot/ardupilot/issues/21946#issuecomment-1277585156

But not the crash....

### CODE REVIEW

File: libraries/AP_Logger/AP_Logger.cpp [hunk 1]
```diff
@@ -1170,6 +1170,113 @@ AP_Logger::log_write_fmt *AP_Logger::msg_fmt_for_name(const char *name, const ch
...
 }
 
+// as above but does not assume pointers will stick around.
+AP_Logger::log_write_fmt *AP_Logger::msg_fmt_for_name_scripting(const char *name, const char *labels, const char *units, const char *mults, const char *fmt)
+{
+    WITH_SEMAPHORE(log_write_fmts_sem);
+    struct log_write_fmt *f;
+    for (f = log_write_fmts; f; f=f->next) {
+        // direct comparison used from scripting where pointer is not maintained
+        if (strcmp(f->name,name) == 0) {
```
Comment (IamPete1) [inline libraries/AP_Logger/AP_Logger.cpp:1180]: Now this is a different function we could keep a separate list of scripting formats, this should significantly speed up scripting logging as we would only have to do one or two string compares.

Comment (IamPete1) [reply libraries/AP_Logger/AP_Logger.cpp:1180]: Could also just up none scripting formats at the end of the linked list....

File: libraries/AP_Logger/AP_Logger.cpp [hunk 2]
```diff
@@ -1170,6 +1170,113 @@ AP_Logger::log_write_fmt *AP_Logger::msg_fmt_for_name(const char *name, const ch
     return f;
 }
 
+// as above but does not assume pointers will stick around.
+AP_Logger::log_write_fmt *AP_Logger::msg_fmt_for_name_scripting(const char *name, const char *labels, const char *units, const char *mults, const char *fmt)
+{
```
[the remark this answers is not in the document]
Comment (IamPete1) [reply libraries/AP_Logger/AP_Logger.cpp:1175]: done

File: libraries/AP_Logger/AP_Logger.cpp [hunk 3]
```diff
@@ -1116,11 +1116,42 @@ AP_Logger::log_write_fmt *AP_Logger::msg_fmt_for_name(const char *name, const ch
...
     f->msg_type = msg_type;
-    f->name = name;
-    f->fmt = fmt;
-    f->labels = labels;
-    f->units = units;
-    f->mults = mults;
+
+    if (copy_strings) {
+        // cannot use pointers to memory that might move, must allocate and copy
+        char* name_copy = (char*)malloc(LS_NAME_SIZE);
```
Comment (tridge) [inline libraries/AP_Logger/AP_Logger.cpp:1122]: nullptr checks

File: libraries/AP_Logger/AP_Logger.cpp [hunk 4]
```diff
@@ -1116,11 +1116,42 @@ AP_Logger::log_write_fmt *AP_Logger::msg_fmt_for_name(const char *name, const ch
...
+        strcpy(name_copy, name);
+
+        char* fmt_copy = (char*)malloc(LS_FORMAT_SIZE);
```
Comment (tridge) [inline libraries/AP_Logger/AP_Logger.cpp:1125]: use strdup, or use single structure and alloc once

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_Logger/AP_Logger.cpp [hunk 5]
```diff
@@ -1116,11 +1116,42 @@ AP_Logger::log_write_fmt *AP_Logger::msg_fmt_for_name(const char *name, const ch
...
-    f->name = name;
-    f->fmt = fmt;
-    f->labels = labels;
-    f->units = units;
-    f->mults = mults;
+
+    if (copy_strings) {
+        // cannot use pointers to memory that might move, must allocate and copy
+        char* name_copy = (char*)malloc(LS_NAME_SIZE);
+        strcpy(name_copy, name);
```
Comment (peterbarker) [inline libraries/AP_Logger/AP_Logger.cpp:1123]:
```suggestion
        f->name = strndup(name)
        if (f->name == nullptr) {
           ...
        }
```

**Comment by IamPete1** [comment]
Moved to a structure and used it in a couple of other places too. Need to re-test.

File: libraries/AP_Logger/AP_Logger.cpp [hunk 6]
```diff
@@ -1116,11 +1116,40 @@ AP_Logger::log_write_fmt *AP_Logger::msg_fmt_for_name(const char *name, const ch
...
+        // cannot use pointers to memory that might move, must allocate and copy
+        struct log_write_fmt_strings *ls_copy = (struct log_write_fmt_strings*)malloc(sizeof(log_write_fmt_strings));
+        if (ls_copy == nullptr) {
+            free(f);
+            return nullptr;
+        }
+
+        strcpy(ls_copy->name, name);
+        strcpy(ls_copy->format, fmt);
+        strcpy(ls_copy->labels, labels);
```
Comment (peterbarker) [inline libraries/AP_Logger/AP_Logger.cpp:1130]:
```suggestion
        strncpy(ls_copy->name, name, sizeof(ls_copy->name));
        strncpy(ls_copy->format, fmt, sizeof(ls_copy->format));
        strncpy(ls_copy->labels, labels, sizeof(ls-copy->labels));
```

... and elsewhere

**Review by peterbarker — DISMISSED** [review]
... because nobody likes accidental memory overwrites, so being extra-special-careful is good.

File: libraries/AP_Logger/AP_Logger.cpp (hunk 6 above)
Comment (IamPete1) [reply libraries/AP_Logger/AP_Logger.cpp]:
```
../../libraries/AP_Logger/AP_Logger.cpp:1128:16: error: 'char* strncpy(char*, const char*, size_t)' specified bound 5 equals destination size [-Werror=stringop-truncation]
 1128 |         strncpy(ls_copy->name, name, sizeof(ls_copy->name));
      |         ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Comment (IamPete1) [reply libraries/AP_Logger/AP_Logger.cpp]: @peterbarker We could either back to `strcpy` or disable the error for this section. Or maybe there is a third solution? What do you think?

Comment (peterbarker) [reply libraries/AP_Logger/AP_Logger.cpp]: `strncpy_noterm`

Comment (IamPete1) [reply libraries/AP_Logger/AP_Logger.cpp]: perfect, thanks!

**Comment by IamPete1** [comment]
Re-tested under valgrind, issue is resolved.

**Review by tridge — APPROVED** [review]

**Pull request merged**